harness/determined/common/experimental/workspace.py

```diff
@@ -0,0 +1,68 @@
+from typing import Optional
+
+from determined.common import api
+from determined.common.api import bindings
+
+
+class Workspace:
+    """A class representing a Workspace object.
+
+    Attributes:
+        id: (str) The ID of the workspace.
+        name: (Mutable, str) The name of the workspace.
+    """
+
+    def __init__(
+        self,
+        session: api.Session,
+        workspace_id: Optional[int] = None,
+        workspace_name: Optional[str] = None,
+    ):
+        """Create a Workspace object.
+
+        Arguments:
+            session: The session to use for API calls.
+            workspace_id: (Optional) ID of the workspace. If None or not passed, the workspace ID
+                will be looked up here at initialization.
+            workspace_name: (Optional) Name of the workspace.
+
+        Note: Only one of workspace_id or workspace_name should be passed.
+        """
+        if (workspace_id is None) == (workspace_name is None):
+            raise ValueError("Workspace must be constructed with either a name or id (not both).")
+
+        if workspace_id is None:
+            assert workspace_name is not None
+            self.id = _get_workspace_id_from_name(session, workspace_name)
+        else:
+            self.id = workspace_id
+        self.name = workspace_name
+        self._session = session
+
+    @classmethod
+    def _from_bindings(
+        cls, workspace_bindings: bindings.v1Workspace, session: api.Session
+    ) -> "Workspace":
+        workspace = cls(session, workspace_id=workspace_bindings.id)
+        workspace._hydrate(workspace_bindings)
+        return workspace
+
+    def _hydrate(self, workspace_bindings: bindings.v1Workspace) -> None:
+        self.id = workspace_bindings.id
+        self.name = workspace_bindings.name
+
+    def reload(self) -> None:
+        resp = bindings.get_GetWorkspace(session=self._session, id=self.id)
+        workspace_bindings = resp.workspace
+
+        self._hydrate(workspace_bindings)
+
+
+def _get_workspace_id_from_name(session: api.Session, name: str) -> int:
```

Comment:
is there a reason why this method is separate from the class? do we envision callers of this method from outside the Workspace class?

Reply:
All other things being equal, I think class methods are harder to reason about than module-level functions. You can be reasonably confident module-level functions have no side effects and you know exactly what the relevant parameters are.

And like I know that classes exist for a reason. Maybe three of them are:
1. Persisting side-effects
2. Cutting down on typing when passing parameters (or just syntactic sugar where `fun(Obj) -> Obj.fun()`)
3. Organizing namespaces

I'm fully in support of (1) when it's needed. Side effects are hard, but if they're convenient to your problem then it's dope to be able to put them into an object. (2)... sometimes I like it, sometimes I don't? Whatever makes the code easier to read and think about, I guess, depending on how much brevity you're gaining for the sacrifice to explicitness. And at least in Python, I think the community has decided that (3) is an antipattern -- modules are this language's preferred organizational tool.

This means that when (1) isn't an issue, I generally prefer to stick helper functions outside the class and directly into the module namespace.

I dunno. This is a preference I picked up somewhere in my last job. If you think it's gross (even just a little) or I've missed other reasons to prefer class-level methods, I'll do my best to talk about it with an open mind.

Reply:
that makes sense. the method doesn't really require persisting of any side-effects, or "state", so it doesn't need to be a method of the class. i was thinking along the lines of 3. organizing namespaces, but you are right that it doesn't seem very pythonic. thanks for the detailed explanation.